src/libraries/System.Private.CoreLib/src/System/Text/UnicodeEncoding.cs

```diff
@@ -1535,7 +1535,7 @@ internal sealed override unsafe int GetCharCount(byte* bytes, int count, Decoder
                     }
 
                     // Valid surrogate pair, add our lastChar (will need 2 chars)
-                    if (chars >= charEnd - 1)
+                    if (chars >= charEnd - 1 || chars == charEnd)
```

Comment:
```suggestion
                    if (charEnd - chars < 2)
```
This should be a bit faster and also easier to understand.

Reply:
Nice optimization.
Fixed.